### PR TITLE
Add support for multiple endpoints using a given handler 

### PR DESCRIPTION
**Pull request by mkatychev**
## Motivation
This PR implements the bullet point below https://github.com/tokio-rs/axum/issues/109:
> * multiple groups in a single path segment

## Solution

* Added `RoutePath` to be used en lieu of `&str` in `route`:

https://github.com/tokio-rs/axum/blob/af3dbeaea604a83eb184594773d0e84a5ab85175/src/routing.rs#L76-L83

```diff
- pub fn route<S, B>(description: &str, service: S) -> Route<S, EmptyRouter<S::Error>>
+ pub fn route<S, B, P>(description: P, service: S) -> Route<S, EmptyRouter<S::Error>>
```

* Moved `route::PathPattern` visibility from `pub(crate)` to `pub`

### CODE REVIEW

**Comment by mkatychev** [comment]
From the guidelines:
> Bug fixes and new features should include tests.

I would include tests but latest `main` is failing quit a few tests for me. 
Happy to move this PR into draft.

File: src/lib.rs [hunk 1]
```diff
@@ -790,9 +790,10 @@ pub mod prelude {
 /// # Panics
 ///
 /// Panics if `description` doesn't start with `/`.
-pub fn route<S, B>(description: &str, service: S) -> Route<S, EmptyRouter<S::Error>>
+pub fn route<S, B, P>(description: P, service: S) -> Route<S, EmptyRouter<S::Error>>
```
Comment (mkatychev) [inline src/lib.rs:793]: should `description` be changed to `path`? Description is usually inert metadata but `P` in this case is used for routing logic.

**Comment by davidpdrsn** [comment]
Thanks for doing this!

However I'm not comfortable merging this. I would like to spend more time researching and comparing other frameworks and then designing an API that solves as much of https://github.com/tokio-rs/axum/issues/109 as possible - while enabling future optimizations to axum's routing we might want to make.

While I think your solution is good it feels a bit like a bandaid and I'm worried it might conflict with a more flexible routing API that we design for https://github.com/tokio-rs/axum/issues/109.

Hope you understand 😊

By the way, a general OSS tip is to first ask a maintainer if they would be open to a change before you start working on a PR. Then you don't run the risk of writing something that doesn't fit with where the authors want to take the project.

**Comment by davidpdrsn** [comment]
> I would include tests but latest main is failing quit a few tests for me.

Oh and `cargo test --all-features` should do the trick (works on my machine).

**Comment by mkatychev** [comment]
> By the way, a general OSS tip is to first ask a maintainer if they would be open to a change before you start working on a PR. Then you don't run the risk of writing something that doesn't fit with where the authors want to take the project.

No worries, I worked on this out of curiosity and to get familiar working with the codebase so no time lost.

**Pull request closed without merging**